### DISCUSSION TITLE
Flows: Filter Using Datachain Values

### DISCUSSION THREAD

**Discussion by MentalGear** [directus/directus#16652] [Archive]
The docs do not show how to access datachain vars from a previous operation, in a flow filter.

Example: The previous flow operation extracts a username. I'd like to use it in a filter.
```
{
    "filter": {
        "username": {
            "_eq": "{{data.get_username}}"
        }
    }
}
```
As a general feedback, I wasted a lot of time debugging flows because the interface for assessing data chain props is not standardized. Sometimes it is data.$last, sometimes just {{ $last }}, etc.

**Comment by br41nslug** [comment]
The docs do mention it (https://docs.directus.io/configuration/flows.html#the-data-chain)
> Once an operation finishes, its data is appended under its \<operationKey>

> In our examples, we are using generic placeholders for operation keys, like \<operationKey>, which might look funny to low-code users. In practice, operation keys will actually have unique and descriptive names, like send_email_7538.

> Remember, $trigger, $accountability, and $last begin with $, but not operationKeys

But you're not wrong the logs view leaves a bit to be desired please upvote/comment this discussion https://github.com/directus/directus/discussions/15870

**Reply by MentalGear** [reply]
Upvoted. Yet my scope was more that the data chain access method should be standardized across usage, e.g. not sometimes with or without data. Still not sure what the right format is for usage within "filter".

**Reply by MentalGear** [reply]
"_eq": "{{data.get_username}}" and  "_eq": "{{ get_username }}" only returns "undefined", how can I access data chain operationKey vars within a "Read Data" `filter` ?

**Reply by br41nslug** [reply]
there should be only 2 very specific situations where you'd use `data.XXXX`.
1. your operation is actually called `data` making `{{ <operationKey>.payload }}` into `{{ data.payload }}`
2. inside of a "Run Script" operation (only if you've stuck with the provided example) where the datachain/context object is passed to the function as an argument named `data`: 
```js
module.exports = async function(data) {
	// do something with "data" or rename the variable and call it something else.
        const username = data.operationKey.get_username;
	return {};
}
```

IMO this is standardized in the form of `<identifier>.dot.notation.accessor`. How you're addressing this data is indeed different when used in different `context`/`programming languages` using it in a Mustache Template (`{{ <identifier>.dot.notation.accessor }}`) or directly in Javascript code (`variableName.<identifier>.dot.notation.accessor`) is not completely avoidable.

You can limit this difference if really wanted by making your "Run Script" code extract the desired operation key directly in the function definition like so (this comes with some character based limitations however):
```js
module.exports = async function({ $last }) {
	// This approach takes a little javascript knowledge
    const user = $last.get_username;
	return {};
}
``` 
Or 
```js
module.exports = async function(data) {
    const complexName = data['
some-complex name];
	// This apoproach takes a little javascript knowledge
    const user = complexName.get_username;
	return {};
}
```

**Reply by MentalGear** [reply]
Thank you for these detailed explications!

You're of course right regarding the technical background of Template interpolation vs JS (vue) var-reference that has the elements wrapped in a data element (Run Script). But TBH though, I think as directus end user only care to have one clear way to do something (especially as flows fail silently and we don't get an error message many times).

I'd be very happy if you could help me figure out my main problem:

<img width="805" alt="Screenshot 2022-11-28 at 18 43 40" src="https://user-images.githubusercontent.com/2837147/204345790-343aec3e-a5c4-4016-93af-61c9668501ed.png">

The username is returned from payload, but I can't transfer it into the next step of the flow:

<img width="832" alt="Screenshot 2022-11-28 at 18 44 08" src="https://user-images.githubusercontent.com/2837147/204345824-c153ede7-becf-4df6-9d97-cba4b0dfc2df.png">

Also tried without $, without data or {} {{}}

Returns in the Log View:
```
{
  "collection": "targets",
  "query": {
    "filter": {
      "username": {
        "_eq": "undefined"
      }
    }
  },
  "permissions": "$full"
}
```

**Reply by br41nslug** [reply]
> I think as directus end user only care to have one clear way to do something (especially as flows fail silently and we don't get an error message many times).

You're not wrong there is definitely room for improvement there.

As for your main problem there:
Because your script operation directly returns the `username` and has the key `get_username` this will result in `'get_username':'valueof data.$trigger.payload.username'` being available on the flow datachain. So if that contains the correct data using a filter like:
```json
{
    "filter": {
      "username": {
        "_eq": "{{ get_username }}",
      }
    }
```
should do the trick (because there is no nesting in the script output just the `<operationKey>` will give you the value it returned)

**Reply by MentalGear** [reply]
Thanks, it's a great note! However turns out my problem in the end was, after trying out so many different variations for template interpolation, I finally simply used the `<opKey>` as you suggested, which didn't work since it was out of sync with the name of the previous operation (that's why I generally prefer using $last).

I'll add a general feedback to the Flows thread regarding `<opKey>` changes. Ideally they would sync automatically if a flow's key/name is changed, across the flow, or otherwise on editing the name of an operation the key shouldn't automatically change. (the first time setting the key based on the name is still good DX)
